frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryTripModelStorage.java

```diff
@@ -0,0 +1,85 @@
+package com.google.tripmeout.frontend.storage;
+
+import com.google.tripmeout.frontend.TripModel;
+import com.google.tripmeout.frontend.error.TripAlreadyExistsException;
+import com.google.tripmeout.frontend.error.TripNotFoundException;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class InMemoryTripModelStorage {
+  Map<String, TripModel> storage = new ConcurrentHashMap<>();
+
+  public void addTrip(TripModel trip) throws TripAlreadyExistsException {
+    synchronized(storage) {
+      if (storage.get(trip.id()) == null) {
+        storage.put(trip.id(), trip);
+      } else {
+        throw new TripAlreadyExistsException("Trip with id: " + trip.id() + 
+            " already exisits");
+      }
+    }
+  }
+
+  public void removeTrip(String tripId) throws TripNotFoundException {
+    TripModel trip = storage.remove(tripId);
+    if (trip == null) {
+      throw new TripNotFoundException("Trip with id: " + tripId + 
+          " not found in storage");
+    }
+  }
+
+  public void updateTripLocation(String tripId, double latitude, double longitude) 
+      throws TripNotFoundException {
+    synchronized(storage) {
+      TripModel trip = storage.get(tripId);
+      if (trip == null) {
+        throw new TripNotFoundException("Trip with id: " + tripId + 
+            " not found in storage");
+      }
+      TripModel updatedTrip = trip.toBuilder()
+          .setLocationLat(latitude)
+          .setLocationLong(longitude)
+          .build();
+
+      storage.put(tripId, updatedTrip);
```

Comment:
You were asking earlier if you should use `ConcurrentHashMap` or `synchronized` blocks in this class. This is a use case where you'll want a `synchronized` block, since you're reading map data, performing an update, and writing it back. However, for this to safely make map updates atomic, _all_ accesses need to be synchronized in the same way (which you're already doing in this change).

As I mentioned above, since you're using synchronization this way, you can drop `ConcurrentHashMap` in favor of a regular old `HashMap` (the former incurs some thread-safety overhead that you don't need).